### PR TITLE
Provide and merge group/node and onms_services from different roles

### LINKED ISSUE #23: Accumulate a list of service test definitions
Each application deployment rule should have a service definition accumulated in a large list assigned to a node when we run the Horizon provisioning role.

### PR DESCRIPTION
**Pull request by mfuhrmann**
feat: Add host/group driven service config
    
Enhance template to create a service by host and group vars. Allow the user to configure the monitored service with
MetaData through Ansible via Host and Group variables. We also install the SNMP agent.
    
Resolve: #21 #23 #24

### CODE REVIEW

File: ansible/host_vars/node1.yml [hunk 1]
```diff
@@ -1,2 +1,6 @@
 ---
 ansible_host: 172.16.238.12
+node_services:
+  HTTP.MyApp:
+    port: "{{ http_port }}"
+    path: "/"
```
Comment (indigo423) [inline ansible/host_vars/node1.yml:6]: What we ideally want is to configure the service test for HTTP.MyApp in role apache2 and not with the node. Do think there is a way to do something like this?

Comment (indigo423) [reply ansible/host_vars/node1.yml:6]: What I would like to do is something like this:

in `roles/apache2/defaults/main.yml`

```yaml
---
onms_services:
  HTTP.MyApp:
    port: "{{ http_port }}"
    path: "/"

```

in `roles/net-snmp/defaults/main.yml`

```
---
onms_services:
  HTTP.MyApp:
    port: "{{ http_port }}"
    path: "/"
```

The behavior is the last `onms_service` list wins and is not merged cause it's the same variable name.

**Comment by mfuhrmann** [comment]
I had to add more services into several places to understand the merge behavior.

The `onms_services` now get merged into the others. We might have to play with the `list_merge` parameter in the combine to get the correct merge behavior when services exist multiple times.

```
TASK [debug] ***************************************************************************************************************
ok: [node1] => {
    "node_services | combine(onms_services) | combine(group_services)": {
        "FTP": null,
        "HTTP.MyApp": {
            "path": "/",
            "port": 80
        },
        "ICMP": null
    }
}
ok: [node2] => {
    "node_services | combine(onms_services) | combine(group_services)": {
        "HTTP": null,
        "HTTP.MyApp": {
            "path": "/",
            "port": 81
        },
        "ICMP": null,
        "TCP": null
    }
}
```

But it feels for me that this approach is getting more and more complex. I think it depends heavily on how you run the playbooks. Here I had to create also one for SNMP which makes sense maybe since I want to roll out SNMP on $myHosts. Before it was included in the `horizon-provision` playbook. By adding the entire 'horizon-provision' role into other application playbooks it seems to work.
I've tried also a subtask in the `horizon-provision` playbook to create only services. But this seems to be stupid, since also the application playbooks would have to make sure, the the requisition, node, service already exist.

But there is still one issue: When I run the snmp playbook I get the expected results. When I then run the Apache playbook, I loose the SNMP service in the node's requisition. It looks like the whole definitions get rewritten. Maybe this is an API call issue. Not sure yet.

I am not taking into account here the service parameters and passing them into the services templates.

I've added here for troubleshooting purposes some debug task:
https://github.com/opennms-forge/ansible-provisioning/pull/26/files#diff-6fc99a14a5f65c6904d46367111c81e699114c6561c5e06489becc24c9f64a22R8-R20

Those can be removed once, we are happy with it.

File: ansible/host_vars/node1.yml (hunk 1 above)
Comment (mfuhrmann) [reply ansible/host_vars/node1.yml]: see https://github.com/opennms-forge/ansible-provisioning/pull/26#issuecomment-1641820665

**Comment by mfuhrmann** [comment]
I've done some tests with `curl`. I have a node with the HTTP-MyApp service and I am adding the SNMP service. It should be the same as we do in Ansible. But it seems to behave different.
In Ansible we only have SNMP or HTTP-MyApp depending on which playbook we've run.

```
mf1045@pe-nb267:~/repos/ansible-provisioning/ansible$ curl -s -u admin:admin http://localhost:8980/opennms/rest/requisitions/stack1-prod | xmllint --format -
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<model-import xmlns="http://xmlns.opennms.org/xsd/config/model-import" date-stamp="2023-07-20T06:46:56.296-04:00" foreign-source="stack1-prod" last-import="2023-07-19T07:15:00.756-04:00">
  <node building="stack1-prod" foreign-id="6332f01f-f80f-4493-8389-dcdc6a7bb-n1" node-label="f369d356bfee">
    <interface descr="disc-if" ip-addr="172.16.238.12" status="1" snmp-primary="P">
      <monitored-service service-name="FTP"/>
      <monitored-service service-name="HTTP-MyApp"/>
      <monitored-service service-name="ICMP"/>
    </interface>
    <category name="prod"/>
    <asset name="serialNumber" value="PF30SFLG"/>
    <asset name="modelNumber" value="20XW005JGE"/>
    <asset name="operatingSystem" value="Ubuntu 22.04.2 LTS"/>
    <meta-data context="node" key="env" value="prod"/>
  </node>
</model-import>


mf1045@pe-nb267:~/repos/ansible-provisioning/ansible$ cat service_snmp.xml 
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<monitored-service xmlns="http://xmlns.opennms.org/xsd/config/model-import" service-name="SNMP"/>


mf1045@pe-nb267:~/repos/ansible-provisioning/ansible$ cat service_http-myapp.xml 
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<monitored-service xmlns="http://xmlns.opennms.org/xsd/config/model-import" service-name="HTTP-MyApp"/>

mf1045@pe-nb267:~/repos/ansible-provisioning/ansible$ curl -v -u admin:admin -H 'Content-type: application/xml' -XPOST -d @service_snmp.xml http://localhost:8980/opennms/rest/requisitions/stack1-prod/nodes/6332f01f-f80f-4493-8389-dcdc6a7bb-n1/interfaces/172.16.238.12/services/
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:8980...
* Connected to localhost (127.0.0.1) port 8980 (#0)
* Server auth using Basic with user 'admin'
> POST /opennms/rest/requisitions/stack1-prod/nodes/6332f01f-f80f-4493-8389-dcdc6a7bb-n1/interfaces/172.16.238.12/services/ HTTP/1.1
> Host: localhost:8980
> Authorization: Basic YWRtaW46YWRtaW4=
> User-Agent: curl/7.85.0
> Accept: */*
> Content-type: application/xml
> Content-Length: 152
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 202 Accepted
< Date: Thu, 20 Jul 2023 10:48:20 GMT
< X-Frame-Options: SAMEORIGIN
< Cache-Control: no-store
< Pragma: no-cache
< Content-Security-Policy: default-src 'none' ; frame-src 'self' ; manifest-src 'self' ; script-src 'self' 'unsafe-inline' 'unsafe-eval'; font-src 'self' https://fonts.googleapis.com  https://fonts.gstatic.com; connect-src 'self' ; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; base-uri 'self' ; form-action 'self' ; img-src 'self' https://tiles.opennms.org  https://*.tile.openstreetmap.org data:
< X-Content-Type-Options: nosniff
< Referrer-Policy: no-referrer
< Permissions-Policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), cross-origin-isolated=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), navigation-override=(), payment=(), picture-in-picture=(), publickey-credentials-get=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()
< Location: http://localhost:8980/opennms/rest/requisitions/stack1-prod/nodes/6332f01f-f80f-4493-8389-dcdc6a7bb-n1/interfaces/172.16.238.12/services/SNMP
< Content-Length: 0
< 
* Connection #0 to host localhost left intact


mf1045@pe-nb267:~/repos/ansible-provisioning/ansible$ curl -s -u admin:admin http://localhost:8980/opennms/rest/requisitions/stack1-prod | xmllint --format -
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<model-import xmlns="http://xmlns.opennms.org/xsd/config/model-import" date-stamp="2023-07-20T06:48:20.239-04:00" foreign-source="stack1-prod" last-import="2023-07-19T07:15:00.756-04:00">
  <node building="stack1-prod" foreign-id="6332f01f-f80f-4493-8389-dcdc6a7bb-n1" node-label="f369d356bfee">
    <interface descr="disc-if" ip-addr="172.16.238.12" status="1" snmp-primary="P">
      <monitored-service service-name="FTP"/>
      <monitored-service service-name="HTTP-MyApp"/>
      <monitored-service service-name="ICMP"/>
      <monitored-service service-name="SNMP"/>
    </interface>
    <category name="prod"/>
    <asset name="serialNumber" value="PF30SFLG"/>
    <asset name="modelNumber" value="20XW005JGE"/>
    <asset name="operatingSystem" value="Ubuntu 22.04.2 LTS"/>
    <meta-data context="node" key="env" value="prod"/>
  </node>
</model-import>
```

**Pull request merged**